[#49144] Fix missing link to Meeting location on index pages

aaron-contreras opened the pull request:
## Bug description
There just wasn't a link being rendered from the get-go, instead it was `html_escaped` plaintext.

## Demo screenshot
![image](https://github.com/opf/openproject/assets/61627014/6bd5d4da-d764-4a96-b9fa-05ccc7204da6)

## Notes
See Bug Work Package: https://community.openproject.org/work_packages/49144

aaron-contreras committed:
Cover buggy behavior of missing link with a spec

Documents the failing behavior with a failing spec which will become
"green" once the expected behavior is implemented (which is that a
link to the Meeting location should be present instead of plaintext
when one is set).

aaron-contreras committed:
Modify rendering of Meeting Location in Table to be an anchor tag

**NOTE**: We don't really need to render a link if there's no location
set. That would result in a blank anchor tag being rendered inside
the `td` element.

aaron-contreras committed:
Refactor: Extract a finder method for the meeting row

This would come in handy and clean up duplicate logic for finding the
relevant meeting row given a `Meeting` instance.

cbliard reviewed (changes requested):
It works, but I'm not sure that's correct. A location can be a url, but it can also be a place description, a city, a room name, etc. So it should be converted to a link only if it's a valid URL.

aaron-contreras committed:
Add specs to ensure invalid URLs are not rendered as links

**NOTE**: A valid meeting location URL is considered to be a URL
that includes the protocol in it. For example:

- `a.com` would not be considered a link to be rendered
- `https://a.com` and `http://a.com` would be considered a renderable
link.

aaron-contreras committed:
Render links to meeting locations only if they are valid URLs

Makes the previously-written specs green ✅.

aaron-contreras commented:
@cbliard Thanks for the catch 👍. Didn't really think of using the Location field as something other than a URL. I've addressed the comment in https://github.com/opf/openproject/pull/13440/commits/7df7ec97a196d56c5a1b794e9814dc8809999736 and https://github.com/opf/openproject/pull/13440/commits/8ac6616f1afdd354252de48690e81d3ed622b27b

oliverguenther commented on modules/meeting/app/components/meetings/row_component.rb:
```diff
@@ -47,7 +47,24 @@ def duration
 
     def location
-      h(model.location)
+      if valid_meeting_location_link?
```
We already use riku for automatically linking to URLs. I think we could just reuse this here? https://github.com/opf/openproject/blob/dev/lib/open_project/text_formatting/filters/autolink_filter.rb#L46

aaron-contreras replied:
Thank you! Determining valid URLs turned out to be more daunting than expected. Addressed here: https://github.com/opf/openproject/pull/13440/commits/bb126714ddbb457401a0e6e2e30cb076339185e7 and a precautionary spec here https://github.com/opf/openproject/pull/13440/commits/6463d44f39bef97125a8dc1bda6dd77b958ff0ce

aaron-contreras committed:
Use `Rinku.auto_link` to handle valid/invalid link generation

Thanks Oliver!

aaron-contreras committed:
Add spec to ensure user-provided HTML is escaped when rendered

aaron-contreras committed:
Merge branch 'dev' into bug/49144-missing-link-in-meetings-location

aaron-contreras committed:
Replace calls to `Rinku.auto_link` with the Rails wrapper method

As provided by the Rinku gem itself, it provides a helper that wraps
`Rinku.auto_link` with some goodies like auto-sanitizing if the
provided String is not marked as `html_safe` and re-marking as
`html_safe` on the result of `Rinku.auto_link` once the string is
acted upon by Rinku.

**NOTE**: The `AutolinkFilter` class has some special nuances in its
usage of css class setting and the `auto_link` helper doesn't really
fit as it would wrap all html_options in a `TagBuilder#tag_options`
and execute, which would end up wrapping the entire HTML and messes up
the Rinku rendering.

oliverguenther reviewed (approved):
Thanks for deep diving into the autolinking 🙇
